### DISCUSSION TITLE
Pydantic error when using RemoteRunnable in Langchain

### DISCUSSION THREAD

**Discussion by LarsAC** [langchain-ai/langserve#478] [Q&A] [answered]
Hello,

my code (simple RAG) runs into an error when I try to use a retriever (PGVector) that I have made available via LangServe:

```
pydantic.v1.error_wrappers.ValidationError: 1 validation error for ConversationalRetrievalChain
retriever
  value is not a valid dict (type=type_error.dict)
```

Relevant part of the client code:
```
from langserve import RemoteRunnable

retriever = RemoteRunnable("http://localhost:8000/retrieval")
[...]
llm_chain = ConversationalRetrievalChain.from_llm(
    llm,
    chain_type="stuff",
    retriever=retriever,
    memory=memory,
    return_source_documents=True,
)
```

It appears to me that the type checking of the RemoteRunnable fails, but I was unable to find a method to make the runnable object appear as the correct type (e.g. VectorStoreRetriever) on the client side.

Versions used:
* langchain==0.1.8
* langchain-core==0.1.25
* langserve[all]==0.0.41
* pydantic==2.6.1

Is this a bug or am I missing something ?

**Comment by eyurtsev** [comment]
Can you share a working example of the the retriever itself? Does the code below work? What does it return?

```python
retriever = RemoteRunnable("http://localhost:8000/retrieval")
retriever.invoke('my query') 
```

**Reply by LarsAC** [reply]
Yes, my test client can invoke the retriever itself as a `RemoteRunnable` exactly as above and it correctly returns a list of Document objects.

**Reply by eyurtsev** [reply, marked as answer]
The type checking needs to be relaxed on ConversationalRetrieverChain. 

For now, what you could do is wrap the RemoteRunnable inside a retriever

```
class MyCustomRetriever(BaseRetriever):
    @abstractmethod
    def _get_relevant_documents(
        self, query: str, *, run_manager: CallbackManagerForRetrieverRun
    ) -> List[Document]:
        """Get documents relevant to a query.
        Args:
            query: String to find relevant documents for
            run_manager: The callbacks handler to use
        Returns:
            List of relevant documents
        """
      invoke the remote runnable here
```

**Reply by LarsAC** [reply]
Great, this does indeed work. Thanks for the workaround !
